llvm/unittests/CAS/LazyAtomicPointerTest.cpp

```diff
@@ -34,26 +35,29 @@ TEST(LazyAtomicPointer, loadOrGenerate) {
   EXPECT_EQ(Ptr.load(), &Value);
 }
 
+#if (LLVM_ENABLE_THREADS)
 TEST(LazyAtomicPointer, BusyState) {
   int Value = 0;
   LazyAtomicPointer<int> Ptr;
   ThreadPool Threads;
 
-  std::mutex BusyStart, BusyEnd;
-  BusyStart.lock();
-  BusyEnd.lock();
+  std::mutex BusyLock, EndLock;
+  std::condition_variable Busy, End;
   Threads.async([&]() {
     Ptr.loadOrGenerate([&]() {
-      BusyStart.unlock();
-      while (!BusyEnd.try_lock()) {
-        // wait till the lock is unlocked.
-      }
+      // Notify busy state.
+      Busy.notify_all();
+      std::unique_lock<std::mutex> LEnd(EndLock);
+      // Wait for end state.
+      End.wait(LEnd);
+      LEnd.unlock();
       return &Value;
     });
   });
 
   // Wait for busy state.
-  std::lock_guard<std::mutex> BusyLockG(BusyStart);
+  std::unique_lock<std::mutex> LBusy(BusyLock);
+  Busy.wait(LBusy);
```

Comment:
Wait can wake up spuriously, so you need some external way to check the state.  I think the phase counter we had before would work here.

Reply:
The test works following:
main thread:     ---> wait for busy state ---> perform test ---> issue end state ---> wait for thread join
update thread: ---> issue busy state ---> wait for end state ---> finish update ---> return

main thread only wakes up in busy state, update thread only wakes up in end state. No sure why we need a phase counter?

Reply:
> No sure why we need a phase counter?

Because condition_variable can return spuriously from `wait`, ie. it can return before anyone actually notifies.

Reply:
I see. I thought that is only the behavior for wait_for. I guess it makes it more and more complicated.

Should we go with the atomic version or conditional variable version?